src/operator/subgraph/subgraph_op.h

```diff
@@ -0,0 +1,181 @@
+  virtual bool UseIncomingLink() const = 0;
+  virtual bool UseOutgoingLink() const = 0;
+  virtual void Reset() {
+  }
+};
+
+using SubgraphSelectPtr = std::shared_ptr<SubgraphSelect>;
+
+/*
+ * This is the interface of the subgraph operator that executes the computation
+ * in the subgraph.
+ */
+class SubgraphOperator {
+  nnvm::Symbol subgraph_sym_;
+public:
+  SubgraphOperator(const nnvm::Symbol &sym) {
+    this->subgraph_sym_ = sym;
+  }
+
+  virtual ~SubgraphOperator() {
+  }
+
+  const nnvm::Symbol &GetSubgraph() const {
+    return subgraph_sym_;
+  }
+
+  virtual void Forward(const OpContext& ctx,
+                       const std::vector<NDArray>& inputs,
+                       const std::vector<OpReqType>& req,
+                       const std::vector<NDArray>& outputs) = 0;
+  virtual void Backward(const OpContext& ctx,
+                        const std::vector<NDArray>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<NDArray>& outputs) = 0;
+};
+
+/*
+ * This provides a set of properties for partitioning a graph into subgraphs,
+ * reconstructing a new graph from the subgraphs and creating a subgraph
+ * operator to execute the subgraph.
+ */
+class SubgraphProperty {
+ public:
+  // the criteria of selecting the subgraph nodes.
+  virtual SubgraphSelectPtr CreateSubgraphSelect() const = 0;
+  // create an nnvm node for a given subgraph. Here users can customize how to
+  // execute the operators in the subgraph.
+  virtual nnvm::NodePtr GetSubgraphNode(const nnvm::Symbol &s) const = 0;
+  // Create a subgraph operator for execution.
+  virtual OpStatePtr CreateSubgraphOperator(const nnvm::Symbol &sym) const = 0;
+  // The type of the subgraph.
+  virtual std::string GetType() const = 0;
+};
+
+using SubgraphPropertyPtr = std::shared_ptr<SubgraphProperty>;
+
+void RegisterSubgraphProperty(SubgraphPropertyPtr property);
+
+/*
+ * This selects nodes for a subgraph that only contains operators
+ * in a given set and it visits nodes via both input and output links.
+ */
+class ContainOpSelect: public SubgraphSelect {
+  std::shared_ptr<const std::unordered_set<std::string>> op_names;
+
+ public:
+  ContainOpSelect(std::shared_ptr<const std::unordered_set<std::string>> op_names) {
+    this->op_names = op_names;
+  }
+
+  virtual bool UseIncomingLink() const {
+    return true;
+  }
+
+  virtual bool UseOutgoingLink() const {
+    return true;
+  }
+
+  virtual bool Select(const nnvm::Node &n,
+                      const std::vector<sg::SimpleNode*> *subgraph_nodes) {
+    return !n.is_variable() && op_names->count(n.op()->name);
+  }
+};
+
+/*
+ * This subgraph property finds a subgraph whose nodes have only operators
+ * within a set. The operators in the subgraph will be executed by _subgraph_op.
+ */
+class SimpleSubgraphProperty: public SubgraphProperty {
+  std::shared_ptr<const std::unordered_set<std::string>> op_names;
```

Comment:
Move this to the end of the class.
https://google.github.io/styleguide/cppguide.html#Declaration_Order